crates/api_common/src/request.rs

```diff
@@ -103,9 +103,16 @@ pub fn generate_post_link_metadata(
     else if federated_thumbnail.is_some() {
       federated_thumbnail
     }
-    // Generate local thumbnail if allowed
-    else if allow_generate_thumbnail {
-      match post.url.or(metadata.opengraph_data.image) {
+    // Generate local thumbnail from metadata if allowed
+    else if allow_generate_thumbnail && !is_image_post {
+      match metadata.opengraph_data.image {
+        Some(url) => generate_pictrs_thumbnail(&url, &context).await.ok(),
+        None => None,
+      }
+    }
+    // Generate local thumbnail from post url if allowed
+    else if allow_generate_thumbnail && is_image_post {
+      match post.url {
```

Comment:
Instead of this change, you should just replace `post.url.or(metadata.opengraph_data.image)` with `post.url.filter(|_| is_image_post).or(metadata.opengraph_data.image)`